web/modules/custom/hoeringsportal_audit_log/src/EventSubscriber/ControllerListener.php

```diff
@@ -0,0 +1,165 @@
+<?php
+
+namespace Drupal\hoeringsportal_audit_log\EventSubscriber;
+
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Drupal\os2web_audit\Service\Logger;
+use Symfony\Component\DependencyInjection\Attribute\Autowire;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Symfony\Component\HttpKernel\Event\ControllerEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Drupal\hoeringsportal_audit_log\Form\SettingsForm;
+use Drupal\node\Entity\Node;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Session\AccountInterface;
+
+/**
+ * Controller listener.
+ */
+final class ControllerListener implements EventSubscriberInterface {
+
+  /**
+   * The module config.
+   *
+   * @var \Drupal\Core\Config\ImmutableConfig
+   */
+  private $moduleConfig;
+
+  /**
+   * Constructor.
+   */
+  public function __construct(
+    ConfigFactoryInterface $configFactory,
+    protected RouteMatchInterface $routeMatch,
+    protected AccountInterface $currentUser,
+    #[Autowire(service: 'os2web_audit.logger')]
+    protected Logger $auditLogger,
+  ) {
+    $this->moduleConfig = $configFactory->get(SettingsForm::SETTINGS);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('config.factory'),
+      $container->get('current_route_match'),
+      $container->get('current_user'),
+      $container->get('os2web_audit.logger')
+    );
+  }
+
+  /**
+   * Audit on route change, by configured routes.
+   *
+   * @param \Symfony\Component\HttpKernel\Event\ControllerEvent $event
+   *   The event to process.
+   */
+  public function onController(ControllerEvent $event) {
+    $pathInfo = $event->getRequest()->getPathInfo();
+    $routeName = $event->getRequest()->attributes->get('_route');
+    $contentTypes = $this->moduleConfig->get('logged_content_types');
+    $loggedRouteNames = $this->moduleConfig->get('logged_route_names');
+    // Initialize content type arrays for view and
+    // edit audit log pages.
+    $view = [];
+    $edit = [];
+
+    $editRoute = 'entity.node.edit_form';
+    $viewRoute = 'entity.node.canonical';
+
+    if ($contentTypes) {
+      // Loop through content types to categorize enabled audit log
+      // pages (view, edit).
+      foreach ($contentTypes as $contentType => $pages) {
```

Comment:
Do we really have to iterate over all the configuration to decide if we need to log? We can easily get the node and check if we're viewing or editing, and then we can make a simple lookup the check if the request must be logged.

``` php
$node = …
$action = [
  'entity.node.edit_form' => 'edit',
  'entity.node.canonical' => 'view',
][$routeName] ?? NULL;
if (1 === ($contentTypes[$node->getType()][$action] ?? NULL)) {
  // Log stuff
  return;
}

// Check if the route name is in the array of hardcoded routes
// from the settings file.
…
```

Checking the routes first (like you actually do) may be even better.